### ISSUE TITLE
Migration of Dedupe Matching Application v2 from Roxy to ml-gradle 

### ISSUE THREAD

**Issue by maeisabelle** [marklogic-community/ml-gradle#207]
An experiment is made to migrate an existing Roxy project to use ml-gradle to test how easy it is to do the migration and how well-documented ml-gradle is. 

The name of the application chosen is **Dedupe Matching Application v2 (DMAv2)**. The progress is pushed under https://project.marklogic.com/repo/projects/APJ/repos/dma-app/browse?at=refs%2Fheads%2Froxy-to-ml-gradle.

**OVERVIEW OF THE APP**
As an overview, it has 4 databases:

- dma-content (the content database of both the app server and rest server)
- dma-modules (the modules database of both the app server and rest server)
- dma-content-test (the content database of both the test server)
- dma-modules-test (the modules database of both the test server)

It has 3 HTTP servers and 1 XDBC server:

- dma (app server)
- dma-rest (rest server)
- dma-test (test server)
- dma-xcc (XDBC)

Both the app server and rest server are using custom rewriters. It also has Roxy unit tests. Inside the app_specific.rb, custom Ruby tasks are defined to extend the default tasks.

Both the UI and the REST API are defined under src, specifically, src/ui for UI and src/rest-api for REST API. Both are deployed in the dma-modules database.

Initial content is deployed to the dma-content database from the data directory.

**MIGRATION STEPS**
 Assuming gradle is already installed and setup properly, here are the steps of the migration from Roxy to ml-gradle

1. Create the build.gradle with the following content.

```
plugins {
  id "net.saliman.properties" version "1.4.6"
  id "com.marklogic.ml-gradle" version "2.7.3"
}
ext {
  mlAppConfig {
    modulePaths = ["src"]
    createTriggersDatabase = false
  }
}

```

_This imports ml-gradle and the Gradle properties plugin. This also sets the modules path to "src" instead of the default "src/main/ml-modules" to minimize code changes. Moreover, since DMAv2 does not need a triggers database, createTriggersDatabase is set to false._

2. Create gradle.properties (and optional environment properties like gradle.local.properties, gradle.dev.properties which is made possible because of the ```net.saliman.properties``` plugin, etc)
```

mlAppName=dma
mlUsername=admin
mlPassword=admin
mlHost=localhost
mlRestPort=30042
mlTestRestPort=30044
mlModulePaths=src
mlNoRestServer=true
mlModulePermissions=rest-admin,read,rest-admin,update,rest-extension-user,execute,dma-role,read,dma-role,update,dma-role,execute
```

_In this properties file, we define the properties we need to override. The full list of supported properties can be found at https://github.com/marklogic-community/ml-gradle/wiki/Property-reference._

3. Create a properties file (build/dma.properties) for defining the resources to be exported from the existing project. Refer to http://localhost:8002/manage/v2 to get the list or refer to ml-config.xml.

```
servers=dma,dma-rest,dma-test,dma-xcc
databases=dma-content,dma-content-test,dma-modules,dma-modules-test
roles=dma-role,dma-role-amped,dma:admin\
  ,dma:app:worker,dma:asti,dma:asti:admin,dma:asti:user:gcpadmin-gov.ph\
  ,dma:team:admin,dma:user:gcpadmin-gov.ph
privilegesExecute=dma:admin:create:admin,dma:admin:create:team,dma:admin:delete:admin\
  ,dma:app:worker:rest,dma:team:admin:create:admin,dma:team:admin:create:user\
  ,dma:team:admin:delete:admin,dma:team:admin:delete:user
privilegesUri=dma:asti,dma:asti:user:gcpadmin-gov.ph
users=dma-user,dma-guest-user,dma-app-worker,gcpadmin@gov.ph
```

4. Run `gradle -PpropertiesFile=build/dma.properties mlExportResources`. _Notice how src/main/ml-config is populated with databases, forests, security, and servers afterwards._

5. Add amps manually under src/main/ml-config/security/amps (**TODO: This should be automated in the next versions of ml-gradle**). You can do this by going to http://localhost:8002/manage/v2/amps, click a list-item then properties then json then save the content to .json file.

6. Rename the exported json files for roles under  src/main/ml-config/security/roles. Make sure the dependent roles are declared after the roles they depend on.

7. Add the password field for all exported json files for users under src/main/ml-config/security/users.

Example:

```

{
  "user-name" : "dma-user",
  "description" : "A user for the dma application",
  **"password" : "dma-password",**
  "role" : [ "dma-role" ]
}


```

8. Run `gradle -i mlDeploy`

9. For deploying the initial content to the content database, update build.gradle and create custom gradle tasks with parameter of type: com.marklogic.gradle.task.MlcpTask. Make sure to import and configure mlcp. This is done by adding the lines below in build.gradle:

```
repositories {
  jcenter()

  // Needed for mlcp dependencies
  maven { url "https://developer.marklogic.com/maven2/" }

  // Needed for hadoop dependencies for mlcp
  maven { url "http://repository.cloudera.com/artifactory/cloudera-repos/" }
}
configurations {
  mlcp {
      resolutionStrategy {
        force "xml-apis:xml-apis:1.4.01"
      }
    }
}
dependencies {
  mlcp "com.marklogic:mlcp:8.0.6.3"
  mlcp "org.apache.commons:commons-csv:1.2"
  mlcp files("lib")
}

ext.dedupeConfigs = ["algorithms", "dictionary", "scores", "transforms", "types", "weights"]

// Define the tasks to import dedupe configs dynamically
dedupeConfigs.each { dedupeConfig ->
    task "import${dedupeConfig}"(type: com.marklogic.gradle.task.MlcpTask) {
        classpath = configurations.mlcp
        command = "IMPORT"
        database = mlAppConfig.contentDatabaseName
        input_file_path = "data/${dedupeConfig}"
        output_permissions = "dma-role,read,dma-role,update,dma-role,execute"
        output_uri_replace = ".*data,''"
    }
}

task importInitialUser(type: com.marklogic.gradle.task.MlcpTask) {
  classpath = configurations.mlcp
  command = "IMPORT"
  database = mlAppConfig.contentDatabaseName
  input_file_path = "data/users/gcpadmin@gov.ph.json"
  output_permissions = "dma-role,read,dma:user:gcpadmin-gov.ph,read,dma:user:gcpadmin-gov.ph,update,dma:user:gcpadmin-gov.ph,execute"
  output_uri_replace = ".*data,''"
}

task importInitialTeam(type: com.marklogic.gradle.task.MlcpTask) {
  classpath = configurations.mlcp
  command = "IMPORT"
  database = mlAppConfig.contentDatabaseName
  input_file_path = "data/team/asti.json"
  output_permissions = "dma-role,read,dma:asti,read,dma:asti,update,dma:asti,execute"
  output_uri_replace = ".*data,''"
}

// Define the task that depends on all deployment of content tasks
task deployContent(dependsOn: [dedupeConfigs.collect{ dedupeConfig -> "import${dedupeConfig}" }, 'importInitialUser', 'importInitialTeam']) {
  doFirst {
    println 'Started deploying content to ' + mlAppConfig.contentDatabaseName
  }
  doLast {
    println 'Finished deploying content'
  }
  group 'Deployment'
}


```

10. Run `gradle -i deployContent`

**NOT YET DONE**

- Migrate other Roxy tasks to Gradle tasks from app_specific.rb (for updating the rest port inside /ui/components.min.js, etc.)
- Loading Roxy unit tests to the test database

**ISSUES ENCOUNTERED**

1. In gradle.properties, you can only specify 1 port but our application is using different ports for the app server, rest server, and xdbc server.

_You can set the mlRestPort to the port of your rest server. As long as the different servers are configured inside src/main/ml-config, ml-gradle will setup your servers accordingly._ 

2. I already configured my rest server (I exported it using ml-gradle using the task `mlExportResources`) inside src/main/ml-config but I'm encountering an issue during deployment related to my rest server.

_Be sure to set `mlNoRestServer=true` inside gradle.properties so it will not override the previously-created rest server you configured inside src/main/ml-config._

3. ml-gradle is not deploying my modules properly.

_Be sure to set `mlModulePaths=src` inside gradle.properties and/or set the lines below inside build.gradle:_

```
ext {
  mlAppConfig {
    modulePaths = ["src"]
  }
}
```

4. I'm not using a trigger database.

_Set this inside build.gradle:_

```
ext {
  mlAppConfig {
    createTriggersDatabase = false
  }
}

```

5. I need to define different properties for different environments:

_Be sure to import the plugin `net.saliman.properties` inside build.gradle and create your environment properties. Be sure to declare it before ml-gradle_

```
plugins {
  _id "net.saliman.properties" version "1.4.6"_
  id "com.marklogic.ml-gradle" version "2.7.3"
}

```

6. I'm having problem deploying the roles I defined under src/main/ml-config: 
`{"errorResponse":{"statusCode":"400", "status":"Bad Request", "messageCode":"MANAGE-INVALIDCONFIG", "message":"MANAGE-INVALIDCONFIG: (err:FOER0000) Invalid server configuration:  Cannot create role"}}`

_Rename the filename of the roles declarations in such a way that the order of the filenames corresponds to their hierarchy. If the role refer to itself, make sure to create a file that declares the role name first, followed by the file with the permissions. Refer to https://github.com/marklogic-community/ml-gradle/tree/master/examples/role-project/src/main/ml-config/security/roles for example and explanation._

7. I'm encountering an error when creating users during deployment:

`{"errorResponse":{"statusCode":"400", "status":"Bad Request", "messageCode":"MANAGE-INVALIDPAYLOAD", "message":"MANAGE-INVALIDPAYLOAD: (err:FOER0000) Payload has errors in structure, content-type or values. User name and password required."}}`

_The mlExportResources does not include the password of the users exported in the files under src/main/ml-config. Be sure to include the password to correct the errors._

_For example:_

```
{
  "user-name" : "dma-user",
  "description" : "A user for the dma application",
  **"password" : "dma-password"**,
  "role" : [ "dma-role" ]
} 


```
8. I can view all my deployed modules in the qconsole but when I go to the homepage of the application, I cannot see anything (I'm getting _404).

_This is a permission error. Be sure to add your default user for the application for the document permission. For DMAv2, the default user is dma-role. This is done by declaring the required document permission for the deployed modules in build.gradle as follows:_

`mlModulePermissions=rest-admin,read,rest-admin,update,rest-extension-user,execute,**dma-role,read,dma-role,update,dma-role,execute**`

**Comment by maeisabelle** [comment]
Hi @damonfeldman  and @rjrudin, 

I cannot update the label for this issue. I think I don't have permission to add labels.

**Comment by dmcassel** [comment]
I just created team-ml-gradle with @rjrudin as the admin. He can choose to add more people to the team or adjust the labels himself.

**Comment by rjrudin** [comment]
This is very helpful. I think almost all of the issues have been addressed for 2.9.0, and I'm now working on exporting amps.

I'm also considering two other things:

1. When migrating Roxy properties, add mlNoRestServer=true by default. 
1. ml-gradle/ml-app-deployer will default to not creating a trigger database; it currently defaults to creating a trigger database.

**Comment by rjrudin** [comment]
Going to open a ticket for setting mlNoRestServer=true by default; already modified ml-app-deployer to not create a triggers database by default.

I believe that captures everything from this ticket. Thanks for putting this together!
